frontends/ios/BitBoxApp/BitBoxApp/Bluetooth.swift

```diff
@@ -297,22 +319,46 @@ class BluetoothManager: NSObject, ObservableObject, CBCentralManagerDelegate, CB
         handleDisconnect()
     }
 
-    func readBlocking(length: Int) -> Data? {
+    struct ReadError: Error {
+        let message: String
+    }
+    
+    func readBlocking(length: Int) throws -> Data {
         if !isConnected() {
-            return nil
+            throw ReadError(message: "not connected")
         }
         print("BLE: wants to read \(length)")
 
+        currentContextLock.lock()
+        guard let ctx = currentContext else {
+            currentContextLock.unlock()
+            throw ReadError(message: "no connection context")
+        }
+        currentContextLock.unlock()
+        
+        let currentID = ctx.identifier;
+
         var data = Data()
 
         // Loop until we've read the required amount of data
         while data.count < length {
-            // Block until BLE reader callback notifies us
-            semaphore.wait()
-            readBufferLock.lock()
-            data.append(readBuffer.prefix(64))
-            readBuffer = readBuffer.advanced(by: 64)
-            readBufferLock.unlock()
+            // Block until BLE reader callback notifies us or the peripheral is disconnected.
+            ctx.semaphore.wait()
+            
+            if !isConnected() {
+                throw ReadError(message: "the peripheral has disconnected while reading")
```

Comment:
probably redundant to the below check, currentContext will be `nil` if not connected.